### PR TITLE
Allow to specify total memory on agent configuration

### PR DESCRIPTION
**Pull request by preetapan**
Fixed up merge conflicts from PR #3778 

Allow to set the total memory of an agent in its configuration file. This
can be used in case the automatic detection doesn't work or in specific
environments when memory overcommit (using swap for example) can be
desirable.

### CODE REVIEW

File: client/fingerprint/memory.go [hunk 1]
```diff
@@ -24,17 +24,25 @@ func NewMemoryFingerprint(logger *log.Logger) Fingerprint {
...
 
 func (f *MemoryFingerprint) Fingerprint(req *cstructs.FingerprintRequest, resp *cstructs.FingerprintResponse) error {
-	memInfo, err := mem.VirtualMemory()
-	if err != nil {
-		f.logger.Printf("[WARN] Error reading memory information: %s", err)
-		return err
+	var totalMemory int
+	cfg := req.Config
+	if cfg.MemoryMB != 0 {
+		totalMemory = cfg.MemoryMB * 1024 * 1024
```
Comment (chelseakomlo) [inline client/fingerprint/memory.go]: Extract into a named constant

File: client/fingerprint/memory.go [hunk 2]
```diff
@@ -24,17 +24,25 @@ func NewMemoryFingerprint(logger *log.Logger) Fingerprint {
 }
 
 func (f *MemoryFingerprint) Fingerprint(req *cstructs.FingerprintRequest, resp *cstructs.FingerprintResponse) error {
-	memInfo, err := mem.VirtualMemory()
-	if err != nil {
-		f.logger.Printf("[WARN] Error reading memory information: %s", err)
-		return err
+	var totalMemory int
+	cfg := req.Config
+	if cfg.MemoryMB != 0 {
```
Comment (chelseakomlo) [inline client/fingerprint/memory.go]: Would writing a unit test for this make sense?

File: client/fingerprint/memory.go [hunk 3]
```diff
@@ -24,17 +24,25 @@ func NewMemoryFingerprint(logger *log.Logger) Fingerprint {
...
 	}
 
-	if memInfo.Total > 0 {
-		resp.AddAttribute("memory.totalbytes", fmt.Sprintf("%d", memInfo.Total))
-
+	if totalMemory > 0 {
+		resp.AddAttribute("memory.totalbytes", fmt.Sprintf("%d", totalMemory))
 		resp.Resources = &structs.Resources{
-			MemoryMB: int(memInfo.Total / 1024 / 1024),
+			MemoryMB: int(totalMemory / 1024 / 1024),
```
Comment (chelseakomlo) [inline client/fingerprint/memory.go]: Named constant here.

**Comment by shantanugadgil** [comment]
@preetapan 

Sorry about digging up old issues, but from: https://github.com/hashicorp/nomad/issues/1480
it seems Docker + swapping has been permanently disabled.

The docs also seem to concur:
https://www.nomadproject.io/docs/drivers/docker.html#memory

So ... is this capability to override the memory value only for exec/raw_exec ?
OR
will this also help while running Docker containers? 😕

**Comment by shantanugadgil** [comment]
One more link to add to the confusion ... https://github.com/hashicorp/nomad/pull/4091

based on the PR by @tbartelmess I think I see than the docker memory settings will **not** get any benefit by setting custom memory value of this PR.

**Pull request merged**